src/helpers/privacy_budget.rs

```diff
@@ -0,0 +1,228 @@
+#![allow(dead_code)]
+
+use crate::error::Res;
+use redis::{transaction, Commands, RedisWrite, ToRedisArgs};
+use rust_elgamal::RistrettoPoint;
+#[derive(Debug, PartialEq, Eq)]
+struct BlindedMatchKey(RistrettoPoint);
+
+impl BlindedMatchKey {
+    pub fn new(key: RistrettoPoint) -> Self {
+        Self(key)
+    }
+}
+
+impl ToRedisArgs for BlindedMatchKey {
+    fn write_redis_args<W>(&self, out: &mut W)
+    where
+        W: ?Sized + RedisWrite,
+    {
+        // let (c0, c1) = self.0.inner();
+        out.write_arg(self.0.compress().as_bytes());
+    }
+}
+
+/// This module implements privacy budget using a redis cache
+struct PrivacyBudget {
+    redis_connection: redis::Connection,
+    total_privacy_budget: u32,
+}
+
+impl PrivacyBudget {
+    fn new(redis_host_name: &str, redis_password: &str, is_tls: bool) -> Res<PrivacyBudget> {
+        //if Redis server needs secure connection
+        let uri_scheme = if is_tls { "rediss" } else { "redis" };
+
+        let redis_conn_url = format!("{}://:{}@{}", uri_scheme, redis_password, redis_host_name);
+
+        let redis_connection = redis::Client::open(redis_conn_url)?.get_connection()?;
+
+        Ok(PrivacyBudget {
+            redis_connection,
+            total_privacy_budget: 100,
+        })
+    }
+
+    /// Input : a vector of ciphertexts alongwith budget to be consumed.
+    /// This atomically increases the budget of all of the blinded match keys which have
+    /// sufficient budget to consume. For those which do not pass, match keys are discarded and not passed to the output.
+    pub fn get_eligible_bmk_with_budget<'a>(
```

Comment:
This function has a side-effect. It updates the Redis database. But the function name begins with "get_". I think that functions which start with "get_" should just retrieve data, not have side-effects. Perhaps we could name this differently to better reflect what it does. I'm thinking something like "consume_budget_and_return_valid". Not an amazing name, feel free to suggest something else.

Reply:
Makes sense. Let me change